### PR TITLE
(GH-371) Enable self-service mode

### LINKED ISSUE #371: Add check for useBackgroundService
ChocolateyGUI needs to look in the config file and see if features/useBackgroundService is there and if it is set to true.

Then allow running install/upgrade without trying to seek admin rights.

### PR DESCRIPTION
**Pull request by ferventcoder**
When using self-service installs, a feature in Chocolatey for Business,
it provides a background agent that actually performs the
installations. Chocolatey itself handles passing through the proper
information to the background service so that installations can occur.
The GUI should notice the feature is switched on and adjust to allow
non-administrators to run certain commands in the GUI without
attempting to elevate.

Supersedes #382
Closes #371

### CODE REVIEW

**Comment by ferventcoder** [comment]
There is still something not quite right here: 

image


Then the connection between the GUI and the subprocess is super erratic. Is this a known issue?

**Comment by ferventcoder** [comment]
I think I found an issue with the background service that didn't bubble up to the GUI at all. 

"[Fatal] Calls to Chocolatey only are allowed. Attempted call to 'C:\vagrant\resources\ChocolateyGiu\Subprocess\ChocolateyGui.Subprocess.exe'. This has been logged for an administrator to review."

**Comment by ferventcoder** [comment]
But that means progress!

**Comment by ferventcoder** [comment]
I'm guessing the fatal error kills the GUI Subprocess.

**Comment by ferventcoder** [comment]
Still getting issues with connections and this as well:

image

**Comment by ferventcoder** [comment]
Looking at the logs it appears things are working appropriately, but something causes it to kill after about 2-3 seconds. Running it with straight choco and it completes successfully.

I wonder if it is because GUI uses `net.tcp` and background mode uses `net.pipe` and something there is incompatible?

**Comment by ferventcoder** [comment]
No, it looks like the agent returns a message that makes the GUI think it is complete.

**Comment by ferventcoder** [comment]
@RichiCoder1 @gep13 appveyor issues don't seem related - it's a cake error.

**Comment by RichiCoder1** [comment]
Looks like it's just a StyleCop error: `ViewModels\LocalSourceViewModel.cs(22,1): error SA1210: Using directives must be ordered alphabetically by the namespaces` 

That run issue is a Xaml issue (that I could have sworn we already fixed.)

**Comment by RichiCoder1** [comment]
`[Fatal]` generally means an exception bubbled all the way up the host and was unhandled. So you can assume that anything with that tag was, well, fatal.

File: Source/ChocolateyGui/ChocolateyGui.csproj [hunk 1]
```diff
@@ -233,7 +234,7 @@
     <Compile Include="Providers\ChocolateyConfigurationProvider.cs" />
     <Compile Include="Providers\ChocolateyCustomSchemeProvider.cs" />
     <Compile Include="Providers\ChocoRequestHandler.cs" />
-    <Compile Include="Providers\ElevationStatusProvider.cs" />
+    <Compile Include="Elevation.cs" />
```
Comment (mwallner) [inline Source/ChocolateyGui/ChocolateyGui.csproj]: this will probably lead to a StyleCop err?

**Comment by mwallner** [comment]
@RichiCoder1 I think we can merge this into develop - will create another intermediate build and do some testing - is that ok with you?

**Comment by ferventcoder** [comment]
As long as all of the xaml is good to go, I'm happy with that - we'll change this from closes though.

**Comment by RichiCoder1** [comment]
I suppose most all the remaining work is in Choco.Lib and the integration layer and those can be fixed in future PRs.

**Comment by ferventcoder** [comment]
There is still some work here to be completed - like the issue with something shutting down the self-service before it completes the install 

And ensuring those values do get set in source options.

**Comment by RichiCoder1** [comment]
@ferventcoder Could you elaborate? I'm not sure I follow.

**Comment by RichiCoder1** [comment]
Good heavens above, rebasing can go die in a fire.

**Comment by ferventcoder** [comment]
Let me just say the rebase was especially painful the last time I did it

**Comment by ferventcoder** [comment]
Source options like skip proxy and allow self-service are not actually saved, so I may have missed a setting somewhere. It looks like they are set, but then they are not.

**Comment by ferventcoder** [comment]
And the self-service aspect still does not actually work for a non-admin - it dies when attempting to do the install. Not sure what causes it just yet.

**Comment by ferventcoder** [comment]
A couple of additional clean up items on this side, but this will work perfectly once the following are released:

* chocolatey-agent v 0.7.0
* chocolatey.extension 1.11.0

I will add those additional items here today.

**Comment by mwallner** [comment]
bump @RichiCoder1 :-)

**Comment by RichiCoder1** [comment]
I believe @ferventcoder was just wrapping up the fixes for this.

**Comment by ferventcoder** [comment]
Was there something left for me here?

**Comment by ferventcoder** [comment]
I know that there was an issue with saving the self-service and bypass proxy settings, but I wasn't sure what I missed.

**Comment by RichiCoder1** [comment]
I suppose the todo items (deploy fixed version of agent and extension) are technically independent of this PR. Have we confirmed that it works w/ agent now?

**Comment by RichiCoder1** [comment]
Once we merge, I'll do follow up PRs to make sure everything is polished.

**Comment by ferventcoder** [comment]
I've confirmed it!

**Comment by ferventcoder** [comment]
I blame you ;)

**Comment by ferventcoder** [comment]
This is so much win right here.

**Comment by mwallner** [comment]
well see x_x - I'll hand out this version to a couple of beta-testers tomorrow ;-)

**Comment by mwallner** [comment]
AppVeyor still blows up because of Cake/StyleCop .. will fix tomorrow - TeamCity is fine.

**Comment by ferventcoder** [comment]
Did you see the only apply background service to non-admins coming in Chocolatey Licensed v1.12?

**Comment by RichiCoder1** [comment]
Ahh, this explains that explosion I saw.

**Pull request merged**